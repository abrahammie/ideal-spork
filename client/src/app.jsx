import React from 'react';
import ReactDOM from 'react-dom';
import moment from 'moment';
import axios from 'axios';
import { Segment, Card } from 'semantic-ui-react';
import { Alert } from './alert.jsx';
import { NewTaskInput } from './newTaskInput.jsx';
import { TaskViewNavigation } from './taskViewNavigation.jsx';

const style = {
  mainDisplay: {
    margin: '40px auto',
    maxWidth: 900,
  },
  segment: {
    backgroundColor: 'rgba(243, 240, 240, 0.75)',
  },
};

class App extends React.Component{
  constructor(props) {
    super(props);
    this.state = {
      inputName: null,
      inputDescription: null,
      inputDate: null,
      formError: false,
      view: 'All Incomplete',
      tasks: [],
    };
    this.submitTask = this.submitTask.bind(this);
    this.changeDate = this.changeDate.bind(this);
    this.handleItemClick = this.handleItemClick.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.completeTask = this.completeTask.bind(this);
    this.deleteTask = this.deleteTask.bind(this);
  };

  componentWillMount() {
    // get tasks from api on mount
    axios.get('http://tasks:3001/api/getTasks')
      .then(res => {
        // convert js date to moment
        let tasksWithMomentDate = res.tasks.map(task => {
          task.date = moment(task.date);
          return task;
        });
        this.setState({ tasks: tasksWithMomentDate },
        () => console.log(this.state.tasks))
      })
      .catch(err => console.log('ERROR',err));
  }

  submitTask() {
    this.setState({ formError: false });
    // check for required fields
    if (this.state.inputName && this.state.inputDate) {
      axios.post('http://tasks/api/add',
        {
          newTask: {
            name: this.state.inputName,
            description: this.state.inputDescription,
            date: this.state.inputDate,
          }
        })
        .then(res => this.setState(
          {
            tasks: res.tasks,
            inputName: null,
            inputDescription: null,
            inputDate: null,
          }
        ))
        .catch(err => console.log(err));
    } else {
      this.setState({ formError: true });
    }
  }

  changeDate(date) {
    this.setState({
      inputDate: date,
    });
  };

  // toggle task view
  handleItemClick(e, { name }) {
    this.setState({ view: name });
  }

  handleChange(e) {
    // passed to newTaskInput form fields
    const obj = {};
    obj[e.target.id] = e.target.value;
    this.setState(obj);
  }

  completeTask(taskId) {
    console.log('complete task called')
    axios.post('http://tasks/api/complete', { taskId })
      .then(res => this.setState({ tasks: res.tasks }))
      .catch(err => console.log(err));
  }

  deleteTask(taskId) {
    axios.delete('http://tasks/api/delete', { taskId })
      .then(res => this.setState({ tasks: res.tasks }))
      .catch(err => console.log(err));
  }

  // check if due today or tomorrow
  isDueSoon(taskItem) {
    const todaysDate = moment();
    const tomorrow = moment().add(1, 'days');
    return !taskItem.completed &&
      (taskItem.date.isSame(todaysDate, 'day') ||
      taskItem.date.isSame(tomorrow, 'day'));
  }

  // check if overdue
  isOverdue(taskItem) {
    return !taskItem.completed &&
      taskItem.date.isBefore(moment(), 'day');
  }

  render() {
    const { tasks, inputDate, inputName, inputDescription, formError, view } = this.state;

    // get totals for tasks due soon or overdue
    const tasksDue = tasks.reduce((obj, task) => {
      if (this.isDueSoon(task)) {
        obj.soon = obj.soon ? ++obj.soon : 1;
      } else if (this.isOverdue(task)) {
        obj.overdue = obj.overdue ? ++obj.overdue : 1;
      };
      return obj;
    },{});

    return (
      <div
        style={style.mainDisplay}>
        <Segment
          style={style.segment}
          raised>
          <Card.Group
            itemsPerRow={3}>
            <Card
              color="grey">
              <Card.Content
                textAlign="center">
                <h1>My Tasks</h1>
              </Card.Content>
            </Card>
            {tasksDue.soon ?
              (<Alert
                label="Due Today Or Tomorrow"
                icon="wait"
                handleItemClick={this.handleItemClick}
                num={tasksDue.soon}
                when="due soon"/>)
              : null}
            {tasksDue.overdue ?
              (<Alert
                label="Overdue"
                icon="warning sign"
                handleItemClick={this.handleItemClick}
                num={tasksDue.overdue}
                when="overdue"/>)
              : null}
          </Card.Group>
          <br/>
          <NewTaskInput
            inputDate={inputDate}
            inputName={inputName}
            inputDescription={inputDescription}
            changeDate={this.changeDate}
            submitTask={this.submitTask}
            handleChange={this.handleChange}
            formError={formError}
          />
          <br/>
          <TaskViewNavigation
            view={view}
            handleItemClick={this.handleItemClick}
            tasks={tasks}
            completeTask={this.completeTask}
            deleteTask={this.deleteTask}
            isDueSoon={this.isDueSoon}
            isOverdue={this.isOverdue}
          />
        </Segment>
      </div>
    );
  };
};

ReactDOM.render(<App/>, document.getElementById('app'));
