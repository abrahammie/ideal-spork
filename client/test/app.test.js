import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
// import toJson from 'enzyme-to-json';
// import { shallow } from 'enzyme';
import { App } from '../src/app.jsx';

describe('Component: App', () => {
  const tasks = ['walk the dog', 'make my bed', 'do the dishes'];

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });

  it('renders a snapshot', () => {
    const tree = renderer.create(<App/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

});