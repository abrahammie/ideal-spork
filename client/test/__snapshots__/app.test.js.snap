// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: App renders a snapshot 1`] = `
<div
  style={
    Object {
      "margin": "40px auto",
      "maxWidth": 900,
    }
  }
>
  <div
    className="ui raised segment"
    style={
      Object {
        "backgroundColor": "rgba(243, 240, 240, 0.75)",
      }
    }
  >
    <div
      className="ui three cards"
    >
      <div
        className="ui grey card"
        href={undefined}
        onClick={[Function]}
      >
        <div
          className="center aligned content"
        >
          <h1>
            My Tasks
          </h1>
        </div>
      </div>
    </div>
    <br />
    <div>
      <div
        className="ui raised segment"
        style={
          Object {
            "backgroundColor": "#f1f5ef",
          }
        }
      >
        <h4
          className="ui header"
          style={
            Object {
              "marginBottom": 5,
            }
          }
        >
          Add a task:
        </h4>
        <form
          action={undefined}
          className="ui tiny form"
          onSubmit={[Function]}
        >
          <div
            className="fields"
          >
            <div
              className="five wide field"
            >
              <label
                htmlFor="inputName"
              >
                Task name
              </label>
              <div
                className="ui input"
                content={undefined}
              >
                <input
                  disabled={undefined}
                  id="inputName"
                  onChange={[Function]}
                  required={undefined}
                  tabIndex={undefined}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="eight wide field"
            >
              <label
                htmlFor="inputDescription"
              >
                Description
              </label>
              <div
                className="ui input"
                content={undefined}
              >
                <input
                  disabled={undefined}
                  id="inputDescription"
                  onChange={[Function]}
                  required={undefined}
                  tabIndex={undefined}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div>
              <div
                className="react-datepicker-wrapper"
              >
                <div
                  className="react-datepicker__input-container"
                >
                  <input
                    autoComplete={undefined}
                    autoFocus={undefined}
                    className=""
                    disabled={false}
                    id={undefined}
                    name={undefined}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Click to select due date"
                    readOnly={undefined}
                    required={undefined}
                    tabIndex={undefined}
                    title={undefined}
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <button
            className="ui yellow mini compact button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            style={
              Object {
                "marginTop": 10,
              }
            }
            tabIndex={undefined}
          >
            Submit
          </button>
          <div
            className="ui mini error message"
          >
            <div
              className="content"
            >
              <p>
                Task name and date are required.
              </p>
            </div>
          </div>
        </form>
      </div>
    </div>
    <br />
    <div
      className="ui segment"
      style={
        Object {
          "minHeight": 330,
        }
      }
    >
      <div
        className="ui grid"
      >
        <div
          className="five wide column"
        >
          <div
            className="ui fluid vertical tabular menu"
          >
            <a
              className="active item"
              onClick={[Function]}
            >
              All Incomplete
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Due Today Or Tomorrow
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Overdue
            </a>
            <a
              className="item"
              onClick={[Function]}
            >
              Completed
            </a>
          </div>
        </div>
        <div
          className="stretched eleven wide column"
        />
      </div>
    </div>
  </div>
</div>
`;
